### DISCUSSION TITLE
How to handle Localized Pathnames not available in all languages

### DISCUSSION THREAD

**Discussion by capsloq** [amannn/next-intl#1009] [Q&A]
Hello,

dealing with localized pathnames. What to do if a piece of content (url) does not exist in a particular language?
e.g.

```
 "/fantasy/historical-fantasy/cowboy-and-cowgirl": {
        "de": "/fantasie/historische-fantasie/cowboy-und-cowgirl",
        "en": "/fantasy/historical-fantasy/cowboy-and-cowgirl",
        "fr": "/fantaisie/fantaisie-historique/cowboy-et-cow-girl",
        "it": "/fantasia/fantasia-storica/cowboy-e-cowgirl",
        "es": "/fantasia/fantasia-historica/vaquero-y-vaquera",
        "pl": "/fantazja/fantazja-historyczna/kowboj-i-kowbojka",
        "cs": "/fantazie/historicka-fantasy/kovboj-a-kovbojka",
        "nl": "/fantasie/historische-fantasie/cowboy-en-cowgirl",
        "sv": "/fantasi/historisk-fantasy/cowboy-och-cowgirl",
        "no": "/fantasi/historisk-fantasy/cowboy-og-cowgirl",
        "fi": "/fantasia/historiallinen-fantasia/cowboy-ja-cowgirl",
        "da": "/fantasi/historisk-fantasy/cowboy-og-cowgirl",
        "ru": "/fantaziya/istoricheskoe-fentezi/kovboy-i-kovboysha"
      },

``` 

Let's say this content does not exist in Italian (it). Simply deleting the entry leads to a build error.
What to do instead? Leave it the value empty?
My goal is  to **not** generate unnecessary alternate hreflang links in the header.

**Comment by capsloq** [comment]
Nobody? Empty strings still lead to an alternate Link in the Header.
Deleting an entry and setting `alternateLinks: false`, gets rid of the build error but typescript is complaining (rightfully).
@amannn Sorry for the ping. I feel like this should be quite a common use case, no? Having content which is not available in every language.

In the docs it says one can write it's own implementation of generating the alternate links. However it feels wrong to first generate a file with all those slugs knowing that some of them are non existing, just to make typescript happy and correcting that mistake inside the middleware afterwards.

**Reply by amannn** [reply]
Hey! This use case is currently unfortunately not supported. The assumption so far is that i18n can be applied to any given page.

However, it seems like there are users with similar use cases, e.g. https://github.com/amannn/next-intl/discussions/955. I think I recently saw another one, but can't find it currently.

I think for content-heavy apps, some users might solve this with routing that is instrumented via a headless CMS—with `next-intl` being used rather for translating app labels that are less related to content.

I can see a point though that supporting this use case with the routing capabilities from `next-intl` might be helpful.

With localized routing, we could accept a config like this:

```js
 "/fantasy/historical-fantasy/cowboy-and-cowgirl": {
        "de": "/fantasie/historische-fantasie/cowboy-und-cowgirl",
        "en": "/fantasy/historical-fantasy/cowboy-and-cowgirl",
        "fr": null
},
```

… where the user explicitly states that a pathname is not available for a certain locale by specifying `null`. That way we can still rely on TypeScript to ensure every route has a localized definition.

The result would be:
1. Can't use navigation APIs to navigate to this route (e.g. via `Link`)
2. An alternate link is disabled for this entry

My concern is mostly that this only works for localized pathnames and not shared pathnames (since there's no pathnames configuration there). Not sure if there's a better option to support this for both routing options. On the other hand, it could be a benefit of using localized pathnames if this is possible, and you could even use this approach if all pathnames are the same (shared pathnames) and opt out of some entries.

So I think the bottom line is that we could support this. If you'd be interested in setting up a PR for this feature, I'd be happy to review it!

The alternative, as you've mentioned, would be turning off alternate links and providing your own implementation (either in the middleware, or via a sitemap).

**Reply by amannn** [reply]
The issue is being tracked here: https://github.com/amannn/next-intl/issues/990
